### ISSUE TITLE
Support HTTP Traceparent header field reading in Go 1.24

### ISSUE THREAD

**Issue by grcevski** [open-telemetry/opentelemetry-go-instrumentation#1318]
Swiss maps have been enabled by default in Go main, which means they will probably become the default in Go 1.24 and break the reading of the HTTP request maps. gRPC is not affected, since we read an array there.

https://go-review.googlesource.com/c/go/+/574516

Go 1.24 is scheduled for release in February (2nd week), we need a fix by then: https://docs.google.com/presentation/d/1jvc0RzrshOOjkeEnsyea2jwjs_zVOcyETG558pqPOK0/mobilepresent?slide=id.g840eaeb4b4_0_0

One possible fix for this is to write a version of the map lookup code which is meant for Swiss maps. Based on some quick code reading I did, this looks more complex now than it used to be. 

Another possible solution we discussed at the Go Auto SIG today is potentially reading the Traceparent field while the headers are being parsed, although this will be slower since we will run a uprobe on every header field. The advantage of this approach is that this works for all Go versions.

**Comment by grcevski** [comment]
I also realized that maybe we simply need to figure out how to work with the swiss maps. To support manual spans we also read a Go map in `fill_partial_tracer_id_from_tracers_map`, which also reads a Go map.

**Comment by MrAlias** [comment]
We need to fully audit the existing probes to determine where we instrument using maps.

We should see if we can find short-term fixes for all these locations if we cannot get support for swiss maps working before their release.

**Comment by MrAlias** [comment]
@grcevski plans to contribute the code from Beyla they used to get a stop-gap solution working for HTTP

**Comment by MrAlias** [comment]
Doing a quick [search for `MAP_BUCKET_TYPE`](https://github.com/search?q=repo%3Aopen-telemetry%2Fopentelemetry-go-instrumentation%20MAP_BUCKET_TYPE&type=code) shows only two locations we are using that definition:

- https://github.com/open-telemetry/opentelemetry-go-instrumentation/blob/d5869d513063a463ebdfa7dcc50eb2f2c4e255ad/internal/pkg/instrumentation/bpf/net/http/server/bpf/probe.bpf.c#L57
- https://github.com/open-telemetry/opentelemetry-go-instrumentation/blob/d5869d513063a463ebdfa7dcc50eb2f2c4e255ad/internal/pkg/instrumentation/bpf/go.opentelemetry.io/otel/traceglobal/bpf/probe.bpf.c#L120

For completeness, it looks like the kafka header is read from a slice: https://github.com/open-telemetry/opentelemetry-go-instrumentation/blob/d5869d513063a463ebdfa7dcc50eb2f2c4e255ad/internal/pkg/instrumentation/bpf/github.com/segmentio/kafka-go/consumer/bpf/probe.bpf.c#L90-L92

**Comment by MrAlias** [comment]
Plan for OTel global instrumentation support:

- Add package constraints to the [uprobes](https://github.com/open-telemetry/opentelemetry-go-instrumentation/blob/d5869d513063a463ebdfa7dcc50eb2f2c4e255ad/internal/pkg/instrumentation/bpf/go.opentelemetry.io/otel/traceglobal/probe.go#L146-L187) that ensure they are not loaded for versions of Go >= 1.24
- Document in [COMPATIBILITY.md](https://github.com/open-telemetry/opentelemetry-go-instrumentation/blob/d5869d513063a463ebdfa7dcc50eb2f2c4e255ad/COMPATIBILITY.md#default-otel-global-providers-compatibility) that ...
   > Versions of `go.opentelemetry.io/otel < v1.33.0` are not supported when using Go >= `1.24`.

**Comment by nikhil-thomas** [comment]
@MrAlias is this issue taken care of by #1573 completely. 
let me know if i can contribute in any way docs/tests etc.

im new here and trying start contributing. 🧑‍💻

**Comment by MrAlias** [comment]
@nikhil-thomas: The work required for the `traceglobal` probe is complete.

@grcevski do you still plan to work on the `net/http` probe?

**Comment by grcevski** [comment]
Yes, I plan to work on the `net/http` probe. Sorry last week was quite busy being the first week back from the holidays, I'll make the PR this week.

**Comment by MrAlias** [comment]
Completed by #1636

We we double-check our implementations when Go 1.24 is released and adjust accordingly.
